### PR TITLE
[FIX] delivery: real cost policy backorder

### PR DESCRIPTION
**Pull request by HANNICHE-Walid**
port of this fix
https://github.com/odoo/odoo/pull/70454

opw-3219711

### CODE REVIEW

**Review by adwid — COMMENTED** [review]
Better to use `git cherry-pick` in such case, so you take the whole commit (diff, author and description)
Also, could you add a test ?

**Comment by HANNICHE-Walid** [comment]
@adwid added the test on enterprise

**Comment by HANNICHE-Walid** [comment]
> @adwid added the test on enterprise

I've closed that PR and add the test on this one

File: addons/delivery/models/stock_picking.py [hunk 1]
```diff
@@ -269,14 +269,13 @@ def _add_delivery_cost_to_so(self):
...
-            else:
-                delivery_line = delivery_lines[0]
-                delivery_line[0].write({
-                    'price_unit': self.carrier_price,
-                    # remove the estimated price from the description
-                    'name': self.carrier_id.with_context(lang=self.partner_id.lang).name,
-                })
+                delivery_lines = [sale_order._create_delivery_line(self.carrier_id, self.carrier_price)]
+            delivery_line = delivery_lines[0]
+            delivery_line[0].write({
```
Comment (adwid) [inline addons/delivery/models/stock_picking.py:274]: Code is a bit weird, why creating a list 
```suggestion
                delivery_lines = sale_order._create_delivery_line(self.carrier_id, self.carrier_price)
            delivery_lines[0].write({
```

File: addons/delivery/tests/test_delivery_cost.py [hunk 2]
```diff
@@ -353,6 +353,70 @@ def test_delivery_cost_gift_card(self):
         self.assertEqual(len(sale_normal_delivery_charges.order_line), 3)
         self.assertEqual(sale_normal_delivery_charges.amount_untaxed, 80.0, "Delivery cost is not Added")
 
+    def test_delivery_real_cost(self):
```
Comment (adwid) [inline addons/delivery/tests/test_delivery_cost.py:356]: Could you add a little explanation/description ?

File: addons/delivery/tests/test_delivery_cost.py [hunk 3]
```diff
@@ -353,6 +353,70 @@ def test_delivery_cost_gift_card(self):
...
+
+        delivery_wizard = Form(self.env['choose.delivery.carrier'].with_context({
+            'default_order_id': so.id,
+            'default_carrier_id': delivery_carrier.id
+        }))
+        delivery_wizard.save().button_confirm()
+
+        delivery_line = so.order_line.filtered(lambda line: line.is_delivery)
+        self.assertEqual(len(delivery_line), 1, "The delivery line is not present on the SO.")
+        self.assertEqual(delivery_line.price_unit, 0, "The delivery cost should be 0.")
```
Comment (adwid) [inline addons/delivery/tests/test_delivery_cost.py:395]:
```suggestion
        self.assertEqual(len(delivery_line), 1)
        self.assertEqual(delivery_line.price_unit, 0, "The invoicing policy of the carrier is set to 'real cost' and that cost is not yet known, hence the 0 value")
```
An assert message should only be written when we need to explain the expected value, not why the assert fails
I mean
```py
self.assertTrue(a, "Should be true")
```
It does not bring anything, the line is clear enough
(You can actually remove most of the below assert messages)

File: addons/delivery/tests/test_delivery_cost.py [hunk 4]
```diff
@@ -353,6 +353,70 @@ def test_delivery_cost_gift_card(self):
...
+        # confirm the back order
+        bo = picking.backorder_ids
+        bo.move_ids[0].quantity_done = 1.0
+        self.assertGreater(bo.shipping_weight, 0.0, "Picking weight should be positive.")
+        bo._action_done()
+        self.assertEqual(bo.carrier_price, 40.0, "Price is not set")
+
+        new_delivery_line = so.order_line.filtered(lambda line: line.is_delivery) - delivery_line
+        self.assertEqual(len(new_delivery_line), 1, "One new line should be created")
+        self.assertEqual(bo.carrier_price, new_delivery_line.price_unit, "The delivery cost is not updated")
```
Comment (adwid) [inline addons/delivery/tests/test_delivery_cost.py:418]: Except if I'm wrong, you are fixing the PU of the SOL (you ensure that the `carrier_price` is correctly propagated)
So, you should reverse the assert arguments: first is what we test, second is what we want
```suggestion
        self.assertEqual(new_delivery_line.price_unit, bo.carrier_price)
```

**Comment by HANNICHE-Walid** [comment]
@adwid thanks for the feedback I've fixed it

**Comment by AntoineVDV** [comment]
Hello, any clue why the original fwp was closed manually?

**Comment by HANNICHE-Walid** [comment]
@AntoineVDV I believe it was just by mistake

**Comment by AntoineVDV** [comment]
No no I just wanted to know if we tried to avoid a specific version on purpose or if the fwp was forgotten.

**Pull request closed without merging**